politeiawww/cmd/piwww/testrun.go

```diff
@@ -446,15 +475,436 @@ func testUserRoutes(admin testUser, minPasswordLength int) error {
+	md, err := createMDFile()
+	if err != nil {
+		return nil, fmt.Errorf("create MD file: %v", err)
+	}
+	files := []pi.File{*md}
+
+	pm := www.ProposalMetadata{
+		Name: "Some proposal name",
+	}
+	if rfp {
+		pm.LinkBy = time.Now().Add(time.Hour * 24 * 30).Unix()
+	}
+	if linkto != "" {
+		pm.LinkTo = linkto
+	}
+	pmb, err := json.Marshal(pm)
+	if err != nil {
+		return nil, err
+	}
+	metadata := []pi.Metadata{
+		{
+			Digest:  hex.EncodeToString(util.Digest(pmb)),
+			Hint:    pi.HintProposalMetadata,
+			Payload: base64.StdEncoding.EncodeToString(pmb),
+		},
+	}
+
+	sig, err := signedMerkleRoot(files, metadata, cfg.Identity)
+	if err != nil {
+		return nil, fmt.Errorf("sign merkle root: %v", err)
+	}
+
+	return &pi.ProposalNew{
+		Files:     files,
+		Metadata:  metadata,
+		PublicKey: hex.EncodeToString(cfg.Identity.Public.Key[:]),
+		Signature: sig,
+	}, nil
+}
+
+// submitNewPropsal submits new proposal and verifies it if given verify bool
+// is true
+func submitNewProposal(verify bool) (string, error) {
+	fmt.Printf("  New proposal\n")
+	pn, err := proposalNewNormal()
+	if err != nil {
+		return "", err
+	}
+	pnr, err := client.ProposalNew(*pn)
+	if err != nil {
+		return "", err
+	}
+
+	// Verify proposal censorship record
+	pr := &pi.ProposalRecord{
+		Files:            pn.Files,
+		Metadata:         pn.Metadata,
+		PublicKey:        pn.PublicKey,
+		Signature:        pn.Signature,
+		CensorshipRecord: pnr.CensorshipRecord,
+	}
+
+	if verify {
+		err = verifyProposal(*pr, publicKey)
+		if err != nil {
+			return "", fmt.Errorf("verify proposal failed: %v", err)
+		}
+	}
+
+	token := pr.CensorshipRecord.Token
+	fmt.Printf("  Proposal submitted: %v\n", token)
+
+	return token, nil
+}
+
+// proposalSetStatus calls proposal set status command
+func proposalSetStatus(token, status, reason string, unvetted bool) error {
+	pssc := proposalStatusSetCmd{
+		Unvetted: unvetted,
+	}
+	pssc.Args.Token = token
+	pssc.Args.Status = status
+	pssc.Args.Reason = reason
+	err := pssc.Execute(nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// proposalCensor censors given proposal
+func proposalCensor(token, reason string, unvetted bool) error {
+	const proposalStatusCensored = "censored"
+	err := proposalSetStatus(token, proposalStatusCensored, reason, unvetted)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// proposalPublic makes given proposal public
+func proposalPublic(token string) error {
+	const proposalStatusPublic = "public"
+	err := proposalSetStatus(token, proposalStatusPublic, "", true)
+	if err != nil {
+		return err
+	}
+	return nil
+}
 
+// proposalAbandon abandons given proposal
+func proposalAbandon(token, reason string) error {
+	const proposalStatusAbandoned = "abandoned"
+	err := proposalSetStatus(token, proposalStatusAbandoned, reason, false)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// proposalEdit edits given proposal
+func proposalEdit(token string, unvetted, random bool) error {
```

Comment:
I would prefer this to be:
```suggestion
func proposalEdit(state pi.PropStateT, token string) error {
```

For two reasons:
1. Listing the state first is how the data model is setup.  Specify the state (which corresponds to a db on the backend) then specify the token (which corresponds to a db entry in that db).
2. Using `pi.PropStateT` makes the function call more descriptive: `proposalEdit(pi.PropStateUnvetted, token)` vs `proposalsEdit(true, token)`

Reply:
Totally agree